---
title: "integrative clustering: clustering analysis using a joint latent variable model"
date: "Created: 2021-01-22; updated: `r Sys.Date()`"
author: 
  - name: "Hua Zou"
    email: "zouhua1@outlook.com"
output: 
  html_notebook:
    codes: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
library(dplyr)
library(tibble)
library(ggplot2)
library(data.table)
library(iCluster)

rm(list = ls())
options(stringsAsFactors = F)
options(future.globals.maxSize = 1000 * 1024^2)

grp <- c("S1", "S2")
grp.col <- c("#6C326C", "#77A2D1")
```

## What is iCluster

> Integrative clustering of multiple genomic data types: Given multiple genomic data types (e.g., copy number, gene expression, DNA methylation) measured in the same set of samples, iCluster fits a regularized latent variable model based clustering that generates an integrated cluster assigment based on joint inference across data types


### load data
```{r}
phen <- fread("../../Result/phenotype/common_survival_data.tsv")
# all_prf <- fread("../../Result/profile/All_filter_merge_profile.tsv")

copyNumber <- fread("../../Result/profile/copyNumber_filter.tsv")
geneExp <- fread("../../Result/profile/geneExp_filter.tsv")
methylation <- fread("../../Result/profile/methylation_filter.tsv")
protein_RPPA <- fread("../../Result/profile/protein_RPPA_filter.tsv")
```

### get datasets for icluster or icluster2 function

* For each matrix, the rows represent samples, and the columns represent genomic features

```{r}
get_profile <- function(dataset = copyNumber,
                        metadata = phen, 
                        tag = "copyNumber"){
  
  # dataset = protein_RPPA
  # metadata = phen
  # tag = "protein_RPPA"
  
  sid <- intersect(phen$Barcode, colnames(dataset)) 
  res <- dataset %>% dplyr::select(c("V1", sid)) %>% 
    mutate(Type=tag) %>%
    mutate(Name=paste(V1, Type, sep = "_")) %>%
    dplyr::select(Name, V1, Type, everything()) %>%
    dplyr::select(-c("V1", "Type")) %>%
    column_to_rownames("Name") %>%
    t()
  return(res)
}

datasets <- list(copyNumber=get_profile(dataset = copyNumber, tag = "copyNumber"),
                 geneExp=get_profile(dataset = geneExp, tag = "geneExp"),
                 methylation=get_profile(dataset = methylation, tag = "methylation"),
                 protein_RPPA=get_profile(dataset = protein_RPPA, tag = "protein_RPPA"))
```


### iCluster

```{r}
fit <- iCluster(datasets = datasets, k=2, lambda=rep(0.2, 4), max.iter = 50, epsilon = 1e-3)
plotiCluster(fit=fit, label=rownames(datasets[[2]]))
compute.pod(fit)

# save(fit, file = "iCluster_fit.RData")
```

* the final sub-clusters label
```{r}
phen_new <- inner_join(phen,
              data.frame(SampleID=rownames(datasets[[2]]), Cluster=fit$clusters), 
             by = c("Barcode"="SampleID")) %>% 
  dplyr::select(Barcode, Cluster, everything()) %>%
  mutate(Cluster=paste0("S", Cluster))

write.csv(phen_new, "../../Result/phenotype/phenotype_cluster_iCluster.csv", row.names = F)

table(phen_new$Cluster)
```


### iCluster2
```{r}
fit2 <- iCluster2(datasets = datasets, k=2, lambda=list(0.2, 0.2, 0.2, 0.2),
                 max.iter = 50, verbose = TRUE)
save(fit2, file = "iCluster2_fit.RData")
```


### version
```{r}
sessionInfo()
```


### Reference 

1. [iCluster package](https://cran.r-project.org/web/packages/iCluster/iCluster.pdf) 

2. [Integrative clustering of multiple genomic data types using a joint latent variable model with application to breast and lung cancer subtype analysis](https://academic.oup.com/bioinformatics/article/25/22/2906/180866)



