---
title: "Differential Gene Expression analysis"
date: "Created: 2021-01-27 Updated: `r Sys.Date()`"
author: 
  - name: "Hua Zou"
    email: "zouhua1@outlook.com"
output: 
  html_notebook:
    codes: hide
---

## Introduction

We perform the Differential analysis between two subtypes via *DEseq2* package on mRNA, *beta2m in lumi* and *lmFit in limma* packages on methylation data. *Mann–Whitney U test/Wilcoxon rank-sum test* or *lmFit* in limma was used to identify the significant copyNumber and RPPA. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
library(dplyr)
library(tibble)
library(ggplot2)
library(data.table)
library(ggrepel)

rm(list = ls())
options(stringsAsFactors = F)
options(future.globals.maxSize = 1000 * 1024^2)

grp <- c("S1", "S2")
grp.col <- c("#6C326C", "#77A2D1")
```


### load data 

* Copy Number: copyNumber

* gene Expression: geneExp

* DNA methylation: methylation

* Reverse Phase Protein Arrays: protein_RPPA

```{r}
phen <- fread("../../Result/phenotype/phenotype_cluster.csv")

copyNumber <- fread("../../Result/profile/copyNumber_filter.tsv")
geneExp <- fread("../../Result/profile/geneExp_filter.tsv")
methylation <- fread("../../Result/profile/methylation_filter.tsv")
protein_RPPA <- fread("../../Result/profile/protein_RPPA_filter.tsv")
```


### Curation Function 

* Storing expression and phenotype data into Expression Set Object

* Default threshold of the two parameters, the absolute value **log2FoldChange** and **Adjust Pvalue** was 1 and 0.05 respectively.

* Displaying Differential features via volcano plot 

```{r}
### Expression Set Object 
get_ExprSet <- function(x=phen, 
                        y=geneExp){
  # x=phen
  # y=geneExp

  sid <- intersect(x$Barcode, colnames(y))
  # phenotype
  phe <- x %>% filter(Barcode%in%sid) %>%
    mutate(Cluster=factor(as.character(Cluster))) %>%
    column_to_rownames("Barcode") 
  
  # profile 
  prf <- y %>% column_to_rownames("V1") %>%
    dplyr::select(all_of(sid))
  
  # determine the right order between profile and phenotype 
  for(i in 1:ncol(prf)){ 
    if (!(colnames(prf)[i] == rownames(phe)[i])) {
      stop(paste0(i, " Wrong"))
    }
  }
  require(convert)
  exprs <- as.matrix(prf)
  adf <-  new("AnnotatedDataFrame", data=phe)
  experimentData <- new("MIAME",
        name="Hua Zou", lab="Hua Lab",
        contact="zouhua1@outlook.com",
        title="KRIC Experiment",
        abstract="The ExpressionSet",
        url="www.zouhua.top",
        other=list(notes="Created from text files"))
  expressionSet <- new("ExpressionSet", exprs=exprs,
                       phenoData=adf, 
                       experimentData=experimentData)
  
  return(expressionSet)
}

### Differential gene expression 
get_DESeq2 <- function(dataset=geneExp_set,
                       group_col=grp,
                       fc=1,
                       pval=0.05){
  
  # dataset=geneExp_set
  # group_col=grp
  # fc=1
  # pval=0.05
  
  pheno <- pData(dataset) 
  edata <- exprs(dataset)
  
  require(DESeq2)
  dds <- DESeqDataSetFromMatrix(
                    countData=round(edata), 
                    colData=pheno,
                    design=~Cluster)
  
  # edata["HP", ] %>% data.frame() %>% setNames("HP") %>%
  #   rownames_to_column("SampleID") %>%
  #   inner_join(pheno %>% rownames_to_column("SampleID"), by = "SampleID") -> a1
  # wilcox.test(HP~Cluster, data = a1)
  # ggplot(a1, aes(x=Cluster, y=log10(HP)))+geom_boxplot()+theme_bw()
    
  
  dds <- DESeq(dds)
  dds_res <- results(dds, pAdjustMethod = "BH", alpha = 0.05) %>% na.omit()
  dds_res <- dds_res[order(dds_res$padj), ]
  dds_res$Enrichment <- NA
  dds_res[which(dds_res$log2FoldChange >= fc & 
                  dds_res$padj < pval), "Enrichment"] <- group_col[2]
  dds_res[which(dds_res$log2FoldChange <= -fc & 
                  dds_res$padj < pval), "Enrichment"] <- group_col[1]
  dds_res[which(abs(dds_res$log2FoldChange) < fc | 
                  dds_res$padj >= pval), "Enrichment"] <- "Nonsignif"
  
  res <- list(fit=dds, diff_res=dds_res)
  
  return(res)
}

### Differential methylation/copyNumber/methylation
get_limma <- function(dataset=methylation_set,
                      group_col=grp,
                      tag="methylation",
                      fc=1,
                      pval=0.05){
  
  # dataset=methylation_set
  # group_col=grp
  # tag="methylation"  
  # fc=1
  # pval=0.05
  
  pheno <- pData(dataset) 
  
  if(tag == "methylation"){
    # transform the beta value into M values via lumi package
    require(lumi)
    edata <- beta2m(exprs(dataset))    
  }else{
    edata <- exprs(dataset)
  }

  
  require(limma)
  design <- model.matrix(~0 + pheno$Cluster)
  rownames(design) <- rownames(pheno)
  colnames(design) <- group_col
  exprSet <- edata  
  
  # show distribution
  boxplot(exprSet)
  plotDensities(exprSet) 
  
  # linear fitting 
  fit <- lmFit(exprSet, design, method = 'ls')
  contrast <- makeContrasts("S1-S2", levels = design) 
  
  # eBayes
  fit2 <- contrasts.fit(fit, contrast)
  fit2 <- eBayes(fit2)
  
  qqt(fit2$t, df = fit2$df.prior + fit2$df.residual, pch = 16, cex = 0.2)
  abline(0, 1)  

  # differential features
  diff_feature <- topTable(fit2, number = Inf, adjust.method = 'BH') %>%
    rownames_to_column("Feature") 
  
  # prof[rownames(prof)%in%"EVI2A", ] %>% data.frame() %>% setNames("EVI2A") %>%
  #   rownames_to_column("SampleID") %>%
  #   inner_join(pheno %>%rownames_to_column("SampleID"), by = "SampleID") -> a1
  # wilcox.test(EVI2A~Cluster, data = a1)
  # ggplot(a1, aes(x=Cluster, y=EVI2A))+geom_boxplot() 
  
  diff_feature[which(diff_feature$logFC >= fc & 
                       diff_feature$adj.P.Val < pval), "Enrichment"] <- group_col[1]
  diff_feature[which(diff_feature$logFC <= -fc & 
                       diff_feature$adj.P.Val < pval), "Enrichment"] <- group_col[2]
  diff_feature[which(abs(diff_feature$logFC) < fc |
                       diff_feature$adj.P.Val >= pval), "Enrichment"] <- "Nonsignif"
  
  diff_res <- diff_feature %>% 
    setNames(c("Feature", "log2FoldChange", "baseMean", "t", 
               "pvalue", "padj", "B", "Enrichment")) %>% 
    dplyr::select(Feature, everything()) %>%
    arrange(padj, log2FoldChange) %>%
    column_to_rownames("Feature")
  
  res <- list(fit=fit2, diff_res=diff_res)
  
  return(res)
}

# Mann–Whitney U test/Wilcoxon rank-sum test
wilcox_fun <- function(x=pData(protein_RPPA_set)%>% data.frame() %>% rownames_to_column("SampleID"), 
                       y=exprs(protein_RPPA_set) %>% data.frame(), 
                       DNAID="SampleID", 
                       GROUP="Cluster",
                       grp1=grp[1], 
                       grp2=grp[2]){
  
  # x=pData(protein_RPPA_set)%>% data.frame() %>% rownames_to_column("SampleID")
  # y=exprs(protein_RPPA_set) %>% data.frame() 
  # DNAID="SampleID"
  # GROUP="Cluster"
  # grp1=grp[1]
  # grp2=grp[2]  

  # determine x with two cols and names are corret
  phe <- x %>% select(c(DNAID, GROUP))
  colnames(phe)[which(colnames(phe) == DNAID)] <- "SampleID"
  colnames(phe)[which(colnames(phe) == GROUP)] <- "Stage"
  if (length(which(colnames(phe)%in%c("SampleID","Stage"))) != 2){
    warning("x without 2 cols: DNAID, GROUP")
  }

  # select groups
  if(length(grp1)){
    phe.cln <- phe %>% filter(Stage%in%c(grp1, grp2)) %>%
      mutate(Stage=factor(Stage, levels = c(grp1, grp2)))
    pr <- c(grp1, grp2)
  } else{
    phe.cln <- phe %>% mutate(Stage=factor(Stage))
    pr <- levels(phe.cln$Stage)
  }

  if (length(levels(phe.cln$Stage)) > 2) {
    stop("The levels of `group` are more than 2")
  }

  # profile
  sid <- intersect(phe.cln$SampleID, colnames(y))
  prf <- y %>% select(sid) %>%
    rownames_to_column("tmp") %>%
    filter(apply(select(., -one_of("tmp")), 1, function(x){sum(x > 0)/length(x)}) > 0.1) %>%
    data.frame() %>% column_to_rownames("tmp") %>%
    t() %>% data.frame()

  # judge no row of profile filter
  if (ncol(prf) == 0) {
    stop("No row of profile to be choosed\n")
  }

  # merge phenotype and profile
  mdat <- inner_join(phe.cln %>% filter(SampleID%in%sid),
                     prf %>% rownames_to_column("SampleID"),
                     by = "SampleID")
  dat.phe <- mdat %>% select(c(1:2))
  dat.prf <- mdat %>% select(-2)

  res <- apply(dat.prf[, -1], 2, function(x, grp){
    dat <- as.numeric(x)
    p <- signif(wilcox.test(dat ~ grp, paired = F)$p.value, 6)
    # median
    md <- signif(median(dat), 4)
    mdn <- signif(tapply(dat, grp, median), 4)
    if ( mdn[1] > mdn[2] & p < 0.05) {
      enrich1 <- pr[1]
    } else if (mdn[1] < mdn[2] & p < 0.05) {
      enrich1 <- pr[2]
    } else if (p > 0.05 | mdn[1] == mdn[2]){
      enrich1 <- "No significance"
    }

    # rank
    rk <- rank(dat)
    rnk <- signif(tapply(rk, grp, mean), 4)
    if ( rnk[1] > rnk[2] & p < 0.05) {
      enrich2 <- pr[1]
    } else if (rnk[1] < rnk[2] & p < 0.05) {
      enrich2 <- pr[2]
    } else if (p > 0.05 | rnk[1] == rnk[2]){
      enrich2 <- "No significance"
    }
    occ <- signif(tapply(dat, grp, function(x){
      round(sum(x > 0)/length(x), 4)}), 4)

    res <- c(p,enrich1,enrich2,occ,md,mdn,rnk)
    return(res)
  }, dat.phe$Stage) %>%
    t(.) %>% data.frame(.) %>%
    rownames_to_column("type") %>%
    varhandle::unfactor(.)

  colnames(res)[2:11] <- c("Pvalue", "Enrich_median", "Enrich_rank",
                           paste0(pr, "_occurence"), "median_all",
                           paste0(pr, "_median"), paste0(pr, "_rank"))
  res$Block <- paste0(pr[1], "_vs_", pr[2])
  number <- as.numeric(table(dat.phe$Stage))
  res$Num <- paste0(pr[1], number[1], "_vs_",
                    pr[2], number[2])
  res.cln <- res %>% select(c(1,12:13, 2:11)) %>%
    mutate(Pvalue=as.numeric(Pvalue)) %>%
    mutate(FDR=p.adjust(Pvalue, method = "BH")) %>%
    arrange(FDR, Pvalue)
  res2 <- res.cln[,c(1:4,14,5:13)]


  # scale profile
  dat.prf.cln <- prf[, -1]
  dat.phe.cln <- dat.phe %>% mutate(Group=ifelse(Stage==pr[1], 0, 1))
  idx <- which(colnames(dat.phe.cln) == "Group")

  # glm result for odd ratios 95%CI
  glmFun <- function(m, n){
    # calculate the glm between profile and group information
    #
    # Args:
    #   m:  result of group information which must to be numeric
    #   n:  taxonomy to be glm
    #
    # Returns:
    #   the glm result of between taxonomy group
    dat.glm <- data.frame(group=m, marker=scale(n, center=T, scale=T))
    model <- summary(glm(group ~ marker, data = dat.glm,
                         family = binomial(link = "logit")))
    res <- signif(exp(model$coefficients["marker",1]) +
                    qnorm(c(0.025,0.5,0.975)) * model$coefficients["marker",1], 2)

    return(res)
  }

  glm_res <- t(apply(dat.prf.cln, 2, function(x, group){
    res <- glmFun(group, as.numeric(x))
    return(res)
  }, group = dat.phe.cln[, idx]))
  Odd <- glm_res %>% data.frame() %>%
    setNames(c("upper", "expected","lower")) %>%
    mutate("Odds Ratio (95% CI)" = paste0(expected, " (", lower, ";", upper, ")"))
  Odd$type <- rownames(glm_res)

  res_merge <- inner_join(res2,
                          Odd[, c(4:5)], by = "type")

  return(res_merge)
}

### volcanfun
volcanofun <- function(datset=geneExp_Diff,
                       group_name=grp,
                       group_col=grp.col,
                       pval=0.05, 
                       fc=1,
                       fc2=4){
    
  # datset=dge_res
  # group_name=grp
  # group_col=grp.col
  # pval=0.01
  # fc=1
  # fc2=4
  
  
  dat <- datset$diff_res %>% data.frame() %>% 
    rownames_to_column("GeneID") %>%
    mutate(color = case_when(log2FoldChange > fc & padj < pval ~ group_name[2],
                             log2FoldChange < -fc & padj < pval ~ group_name[1],
                             padj > pval | abs(log2FoldChange) < fc ~ "Nonsignif")) %>%
    mutate(color=factor(color, levels = c(group_name, "Nonsignif")))
  print(table(dat$color))
  
  dat.signif <- subset(dat, padj < pval & abs(log2FoldChange) > fc2)
  print(table(dat.signif$color))
  
  group_col_new <- c(group_col, "darkgray")
  group_name_new <- levels(dat$color)
  
  xlabel <- paste0("log2(", paste(group_name, collapse="/"), ")")
  
  # Make a basic ggplot2 object with x-y values
  pl <- ggplot(dat, aes(x = log2FoldChange, y = -log10(padj), color = color))+ 
          geom_point(size = 0.8, alpha = 0.8, shape = 16)+
          scale_color_manual(name = "Enrichment",
                             values = group_col_new,
                             labels = group_name_new)+
          xlab(xlabel)+ 
          ylab(expression(-log[10]("adjusted p-value")))+ 
          geom_hline(yintercept=-log10(pval), alpha=.8, linetype=2, size=.7)+
          geom_vline(xintercept=fc, alpha=.8, linetype=2, size=.7)+
          geom_vline(xintercept=-fc, alpha=.8, linetype=2, size=.7)+ 
          geom_text_repel(data = dat.signif,
                          aes(label = GeneID),
                          size = 4,
                          max.overlaps = getOption("ggrepel.max.overlaps", default = 80),
                          segment.linetype = 1,
                          segment.curvature = -1e-20,
                          box.padding = unit(0.35, "lines"),
                          point.padding = unit(0.3, "lines"),
                          arrow = arrow(length = unit(0.005, "npc")),
                          color = "white",     # text color 
                          bg.color = "grey30", # shadow color
                          bg.r = 0.15)+
          annotate("text", x=min(dat$log2FoldChange), y=-log10(pval), label=pval, size=6, color="red")+
          annotate("text", x=fc, y=0, label=fc, size=6, color="red")+
          annotate("text", x=-fc, y=0, label=-fc, size=6, color="red")+
          scale_y_continuous(trans = "log1p")+
          theme_bw()+ 
          theme(axis.title = element_text(face = 'bold',color = 'black',size = 14),
                axis.text = element_text(color = 'black',size = 10),
                text = element_text(size = 8, color = "black", family="serif"),
                panel.grid = element_blank(),
                legend.position = 'right',
                legend.key.height = unit(0.6,'cm'),
                legend.text = element_text(face = "bold", color = 'black',size = 10),
                strip.text = element_text(face = "bold", size = 14))
  
  res <- list(dge=dat, pl=pl)
  
  return(res)
}
```
 

### copyNumber
```{r}
# expressonSet
copyNumber_set <- get_ExprSet(x = phen, y = copyNumber)

if(!dir.exists("../../Result/Differential")){
  dir.create("../../Result/Differential", recursive = T)
}
saveRDS(copyNumber_set, "../../Result/Differential/copyNumber_ExpressionSet.RDS")

# differetial feature
copyNumber_Diff <- get_limma(dataset = copyNumber_set, group_col = grp, tag = "copyNumber")
saveRDS(copyNumber_Diff, "../../Result/Differential/copyNumber_Diff.RDS")

# volcano
copyNumber_vol <- volcanofun(datset=copyNumber_Diff,
                      group_name=grp,
                      group_col=grp.col,
                      pval=0.05, 
                      fc=1,
                      fc2=4)
copyNumber_vol$pl


if(!dir.exists("../../Result/figure/Differential")){
  dir.create("../../Result/figure/Differential", recursive = T)
}

ggsave(filename = "../../Result/figure/Differential/copyNumber_volcano.pdf", 
       copyNumber_vol$pl, width = 8, height = 5, dpi = 600)

saveRDS(copyNumber_vol, "../../Result/Differential/copyNumber_volcano.RDS")
```


### geneExp
```{r}
# expressonSet
geneExp_set <- get_ExprSet(x = phen, y = geneExp)

if(!dir.exists("../../Result/Differential")){
  dir.create("../../Result/Differential", recursive = T)
}
saveRDS(geneExp_set, "../../Result/Differential/geneExp_ExpressionSet.RDS")

# differetial feature
geneExp_Diff <- get_DESeq2(dataset = geneExp_set, group_col = grp)
saveRDS(geneExp_Diff, "../../Result/Differential/geneExp_Diff.RDS")

# volcano
geneExp_vol <- volcanofun(datset=geneExp_Diff,
                      group_name=grp,
                      group_col=grp.col,
                      pval=0.05, 
                      fc=1,
                      fc2=4)
geneExp_vol$pl


if(!dir.exists("../../Result/figure/Differential")){
  dir.create("../../Result/figure/Differential", recursive = T)
}

ggsave(filename = "../../Result/figure/Differential/geneExp_volcano.pdf", 
       geneExp_vol$pl, width = 8, height = 5, dpi = 600)

saveRDS(geneExp_vol, "../../Result/Differential/geneExp_volcano.RDS")
```


### DNA methylation
```{r}
# expressonSet
methylation_set <- get_ExprSet(x = phen, y = methylation)

if(!dir.exists("../../Result/Differential")){
  dir.create("../../Result/Differential", recursive = T)
}
saveRDS(methylation_set, "../../Result/Differential/methylation_ExpressionSet.RDS")

# differetial feature
methylation_Diff <- get_limma(dataset = methylation_set, group_col = grp, tag = "methylation")
saveRDS(methylation_Diff, "../../Result/Differential/methylation_Diff.RDS")

# volcano
methylation_vol <- volcanofun(datset=methylation_Diff,
                      group_name=grp,
                      group_col=grp.col,
                      pval=0.05, 
                      fc=1,
                      fc2=1)
methylation_vol$pl


if(!dir.exists("../../Result/figure/Differential")){
  dir.create("../../Result/figure/Differential", recursive = T)
}

ggsave(filename = "../../Result/figure/Differential/methylation_volcano.pdf", 
       methylation_vol$pl, width = 8, height = 5, dpi = 600)

saveRDS(methylation_vol, "../../Result/Differential/methylation_volcano.RDS")
```

### protein_RPPA
```{r}
# expressonSet
protein_RPPA_set <- get_ExprSet(x = phen, y = protein_RPPA)

if(!dir.exists("../../Result/Differential")){
  dir.create("../../Result/Differential", recursive = T)
}
saveRDS(protein_RPPA_set, "../../Result/Differential/protein_RPPA_ExpressionSet.RDS")

# differetial feature
protein_RPPA_Diff <- get_limma(dataset = protein_RPPA_set, group_col = grp, tag = "protein_RPPA")
saveRDS(protein_RPPA_Diff, "../../Result/Differential/protein_RPPA_Diff.RDS")

# volcano
protein_RPPA_vol <- volcanofun(datset=protein_RPPA_Diff,
                      group_name=grp,
                      group_col=grp.col,
                      pval=0.05, 
                      fc=1,
                      fc2=1)
protein_RPPA_vol$pl


if(!dir.exists("../../Result/figure/Differential")){
  dir.create("../../Result/figure/Differential", recursive = T)
}

ggsave(filename = "../../Result/figure/Differential/protein_RPPA_volcano.pdf", 
       protein_RPPA_vol$pl, width = 8, height = 5, dpi = 600)

saveRDS(protein_RPPA_vol, "../../Result/Differential/protein_RPPA_volcano.RDS")
```


### version
```{r}
sessionInfo()
```


### Reference 

1. [lumi: transform the beta value into M values](https://rdrr.io/bioc/lumi/man/beta2m.html)


2. [R包limma作差异基因分析](https://cloud.tencent.com/developer/article/1667505)

