---
title: "Classification: extreme gradient boosting(XGBoost)"
date: "Created: 2021-01-20; updated: `r Sys.Date()`"
author: 
  - name: "Hua Zou"
    email: "zouhua1@outlook.com"
output: 
  html_notebook:
    codes: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
library(dplyr)
library(tibble)
library(data.table)
library(caret)
library(pROC)
library(xgboost)

rm(list = ls())
options(stringsAsFactors = F)
options(future.globals.maxSize = 1000 * 1024^2)

grp <- c("S1", "S2")
grp.col <- c("#6C326C", "#77A2D1")
```

## Introduction

> Boosting is a technique in machine learning that has been shown to produce models with high predictive accuracy.
One of the most common ways to implement boosting in practice is to use XGBoost, short for “extreme gradient boosting.”


### load data 

* 4-omics data 

* Copy Number: copyNumber

* gene Expression: geneExp

* DNA methylation: methylation

* Reverse Phase Protein Arrays: protein_RPPA

```{r}
phen <- fread("../../Result/phenotype/phenotype_cluster.csv")
all_norm <- fread("../../Result/classification/all_top_feature_norm.csv")
copyNumber_norm <- fread("../../Result/classification/copyNumber_top_feature_norm.csv")
geneExp_norm <- fread("../../Result/classification/geneExp_top_feature_norm.csv")
methylation_norm <- fread("../../Result/classification/methylation_top_feature_norm.csv")
protein_RPPA_norm <- fread("../../Result/classification/protein_RPPA_top_feature_norm.csv")
```


### Data Splitting & build model based on XGBoost

The alternative parameters for dataset and algorithm in this function 

```{r}
build_XGBoost <- function(dataset=protein_RPPA_norm){
  
  # dataset=protein_RPPA_norm
  
  mdat <- dataset %>% #mutate(Cluster=ifelse(Cluster == grp[1], 0, 1))%>% 
    mutate(Cluster=factor(Cluster, levels = grp))%>%
    column_to_rownames("V1")   
  
  # data partition
  set.seed(123)
  index <- createDataPartition(mdat$Cluster, p = 0.8, list = F)
  trainData_all <- mdat[index, ] 
  testData_all <- mdat[-index, ]
  trainData <- trainData_all %>% dplyr::select(colnames(phen)[2], colnames(mdat)[-c(1:11)]) 
  testData <- testData_all %>% dplyr::select(colnames(phen)[2], colnames(mdat)[-c(1:11)])  
  
  X_train <- xgb.DMatrix(as.matrix(trainData %>% select(-Cluster)))
  y_train <- trainData$Cluster
  X_test <- xgb.DMatrix(as.matrix(testData %>% select(-Cluster)))
  y_test <- testData$Cluster
  

  xgb_trcontrol <- trainControl(
                           method = "repeatedcv", 
                           number = 10,
                           repeats = 3,
                           search = "grid",
                           classProbs = TRUE,
                           allowParallel = TRUE, 
                           returnData = FALSE,
                           verboseIter = FALSE)
  
  xgbGrid <- expand.grid(nrounds = c(100, 200),
                         max_depth = seq(10, 25, 5),
                         colsample_bytree = seq(0.5, 0.9, length.out = 5),
                         eta = 0.1,
                         gamma = 0,
                         min_child_weight = 1,
                         subsample = 1)
  
  set.seed(123) 
  xgb_model <- train(X_train, y_train,  
                     trControl = xgb_trcontrol,
                     tuneGrid = xgbGrid,
                     method = "xgbTree",
                     verbose = FALSE)

  pred <- predict(xgb_model, newdata=X_test)
  print(confusionMatrix(pred, y_test))
  
  
  res <- list(tData_all=testData_all, tData=testData, model=xgb_model)
  return(res)
}


## ROC/AUC
ROC_fun <- function(model_fit=protein_RPPA_ROC_xgboost){
  
  # model_fit=protein_RPPA_model_xgboost
  
  fit <- model_fit$model
  tData <- model_fit$tData
  
  rocobj <- roc(tData$Cluster, predict(fit, newdata = tData[, -1], type = "prob")[, grp[1]])
  auc <- round(auc(tData$Cluster, predict(fit, newdata = tData[, -1], type = "prob")[, grp[1]]), 4)
  
  plroc <- ggroc(rocobj, color = "black", linetype = 1, size = 1, alpha = 1, legacy.axes = T)+
                  geom_abline(intercept = 0, slope = 1, color="grey", size = 1, linetype=1)+
                labs(x = "False Positive Rate (1 - Specificity)",
                     y = "True Positive Rate (Sensivity or Recall)")+
                annotate("text",x = .75, y = .25,label=paste("AUC =", auc),
                         size = 5, family="serif")+
                coord_cartesian(xlim = c(0, 1), ylim = c(0, 1))+
                theme_bw()+
                theme(panel.background = element_rect(fill = 'transparent'),
                      axis.ticks.length = unit(0.4, "lines"), 
                      axis.ticks = element_line(color='black'),
                      axis.line = element_line(size=.5, colour = "black"),
                      axis.title = element_text(colour='black', size=12,face = "bold"),
                      axis.text = element_text(colour='black',size=10,face = "bold"),
                      text = element_text(size=8, color="black", family="serif"))
  return(plroc)  
}

### Evaluation
Evaluation_fun <- function(model_fit=protein_RPPA_model_xgboost){
  
  # model_fit=protein_RPPA_model_xgboost
  
  fit <- model_fit$model
  tData <- model_fit$tData
  tData_all <- model_fit$tData_all
  
  pred <- predict(fit, newdata=tData)
  
  # test Data
  testData_new <- data.frame(SampleID=rownames(tData),
                             Predict_Cluster=pred) %>%
    inner_join(tData_all %>% rownames_to_column("SampleID"),
               by = "SampleID") %>%
    dplyr::select(SampleID, Cluster, Predict_Cluster, OS, OS.Time) %>%
    na.omit()

  # Concordance index
  Concordance_index_fun <- function(dataset=testData_new, 
                                    group_name="Cluster"){
    # dataset=testData_new
    # group_name="Cluster"
    
    require(survcomp)
    colnames(dataset)[which(colnames(dataset) == group_name)] <- "Group"
    
    ci_res <- with(dataset, concordance.index(x=Group, surv.time=OS.Time, surv.event=OS, 
                      method="noether"))  
    
    res <- data.frame(Type=group_name,
                       C_index=signif(ci_res$c.index, 3),
                       pvalue=ci_res$p.value,
                       se=signif(ci_res$se, 3),
                       odd=with(ci_res, paste0("[", signif(lower, 3), ", ", signif(upper, 3), "]")))
    
    return(res)
  }
  Cluster_ci <- Concordance_index_fun(dataset=testData_new, group_name="Cluster")
  Predict_Cluster_ci <- Concordance_index_fun(dataset=testData_new, group_name="Predict_Cluster")
  ci_res <- rbind(Cluster_ci, Predict_Cluster_ci)
  #DT::datatable(ci_res)

  # Log-rank p-value of Cox-PH regression
  Log_rank_fun <- function(dataset=testData_new, 
                           group_name="Cluster"){
    # dataset=testData_new
    # group_name="Cluster"
    
    require(survival)
    require(survminer)
    colnames(dataset)[which(colnames(dataset) == group_name)] <- "Group"
    cox <- coxph(Surv(OS.Time, OS) ~ Group, data = dataset) 
    tmp <- summary(cox)
    tmp.wald <- data.frame(t(tmp$waldtest)) %>%
          setNames(c("Wald_test", "Wald_df", "Wald_pvlaue"))
    tmp.lg <- data.frame(t(tmp$logtest)) %>%
          setNames(c("lg_rank", "lg_rank_df", "lg_rank_pvlaue"))
    tmp.total <- cbind(tmp.wald, tmp.lg) 
      
    pvalue <- paste(paste0("Log-Rank P=", signif(tmp.lg$lg_rank_pvlaue, 3)), 
                    paste0("Cox P=", signif(tmp.wald$Wald_pvlaue, 3)), sep = "\n")
    factors <- levels(dataset$Group)
    surv_fit <- survfit(Surv(OS.Time, OS) ~ Group, data = dataset)
    
    sur_pl <- ggsurvplot(surv_fit,
               data = dataset,  
               surv.median.line = "hv",
               add.all = TRUE,
               palette = "aaas",
               risk.table = TRUE, 
               xlab = "Follow up time(Years)", 
               legend = c(0.8, 0.2), 
               legend.title = "",
               legend.labs = c("all", factors), 
               break.x.by = 2,
               font.legend = c(10, "italic"),
               ggtheme = theme_bw())
    sur_pl$plot <- sur_pl$plot + 
               annotate("text", x=3, y=0.2, label=pvalue)
      
    pl <- cowplot::plot_grid(sur_pl$plot, sur_pl$table, 
                             ncol = 1, align = "hv", 
                             rel_heights = c(2, 1))
      
    pval <- data.frame(Type=group_name,
                             tmp.total)
    res <- list(plot=pl, pvalue=pval)
    return(res)
  }
  
  Cluster_surv <- Log_rank_fun(dataset=testData_new, group_name="Cluster")
  Predict_Cluster_surv <- Log_rank_fun(dataset=testData_new, group_name="Predict_Cluster")
  surv_res <- rbind(Cluster_surv$pvalue, Predict_Cluster_surv$pvalue)
  #DT::datatable(surv_res)

  surv_pl <- cowplot::plot_grid(Cluster_surv$plot, Predict_Cluster_surv$plot,
                     ncol = 2, align = "hv", labels = c("Actual", "Predict"))
    
  # Brier score
  Brier_score_fun <- function(dataset=testData_new, 
                              group_name="Cluster"){
    # dataset=testData_new
    # group_name="Cluster"
    
    require(survcomp)
    colnames(dataset)[which(colnames(dataset) == group_name)] <- "Group"
    
    dat <- dataset %>% dplyr::select(c("Group", "OS", "OS.Time")) %>%
      setNames(c("score", "event", "time"))
    bsc <- sbrier.score2proba(data.tr=dat, data.ts=dat, method="cox")  
    
    res <- list(res_dat=dat, res_bsc=bsc)
    return(res)
  }
  
  Cluster_bsc <- Brier_score_fun(dataset=testData_new, group_name="Cluster")
  Predict_Cluster_bsc <- Brier_score_fun(dataset=testData_new, group_name="Predict_Cluster")
  
  # if(!all(Cluster_bsc$res_dat$time == Predict_Cluster_bsc$res_dat$time)) {
  #   stop("the two vector of BSCs must be computed for the same points in time!") 
  # }
  # ibsc.comp(bsc1=Cluster_bsc$res_bsc$bsc, 
  #           bsc2=Predict_Cluster_bsc$res_bsc$bsc, time=Cluster_bsc$res_dat$time)  
  
  bsc_res <- data.frame(Type=c("Actual", "Predict"),
                        Brier_score=c(Cluster_bsc$res_bsc$bsc.integrated,
                                      Predict_Cluster_bsc$res_bsc$bsc.integrated))
  
  res <- list(ci_res=ci_res, surv_res=surv_res, bsc_res=bsc_res)
  return(res)
}
```


### 4 omics-data
```{r}
all_model_xgboost <- build_XGBoost(dataset = all_norm)

# ROC/AUC 
all_ROC_xgboost <- ROC_fun(model_fit = all_model_xgboost)
all_ROC_xgboost

# evaluation index 
all_Evaluation_xgboost <- Evaluation_fun(model_fit = all_model_xgboost)
all_Evaluation_xgboost

# saving model 
if(!dir.exists("../../Result/classification/model")){
  dir.create("../../Result/classification/model", recursive = T)
}
save(all_model_xgboost, all_ROC_xgboost, all_Evaluation_xgboost, 
     file = "../../Result/classification/model/all_xgboost.RData")
```


### copyNumber
```{r}
copyNumber_model_xgboost <- build_XGBoost(dataset = copyNumber_norm)

# ROC/AUC 
copyNumber_ROC_xgboost <- ROC_fun(model_fit = copyNumber_model_xgboost)
copyNumber_ROC_xgboost

# evaluation index 
copyNumber_Evaluation_xgboost <- Evaluation_fun(model_fit = copyNumber_model_xgboost)
copyNumber_Evaluation_xgboost

# saving model 
if(!dir.exists("../../Result/classification/model")){
  dir.create("../../Result/classification/model", recursive = T)
}
save(copyNumber_model_xgboost, copyNumber_ROC_xgboost, copyNumber_Evaluation_xgboost, 
     file = "../../Result/classification/model/copyNumber_xgboost.RData")
```


### geneExp
```{r}
geneExp_model_xgboost <- build_XGBoost(dataset = geneExp_norm)

# ROC/AUC 
geneExp_ROC_xgboost <- ROC_fun(model_fit = geneExp_model_xgboost)
geneExp_ROC_xgboost

# evaluation index 
geneExp_Evaluation_xgboost <- Evaluation_fun(model_fit = geneExp_model_xgboost)
geneExp_Evaluation_xgboost

# saving model 
if(!dir.exists("../../Result/classification/model")){
  dir.create("../../Result/classification/model", recursive = T)
}
save(geneExp_model_xgboost, geneExp_ROC_xgboost, geneExp_Evaluation_xgboost, 
     file = "../../Result/classification/model/geneExp_xgboost.RData")
```


### methylation
```{r}
methylation_model_xgboost <- build_XGBoost(dataset = methylation_norm)

# ROC/AUC 
methylation_ROC_xgboost <- ROC_fun(model_fit = methylation_model_xgboost)
methylation_ROC_xgboost

# evaluation index 
methylation_Evaluation_xgboost <- Evaluation_fun(model_fit = methylation_model_xgboost)
methylation_Evaluation_xgboost

# saving model 
if(!dir.exists("../../Result/classification/model")){
  dir.create("../../Result/classification/model", recursive = T)
}
save(methylation_model_xgboost, methylation_ROC_xgboost, methylation_Evaluation_xgboost, 
     file = "../../Result/classification/model/methylation_xgboost.RData")
```


### protein_RPPA
```{r}
protein_RPPA_model_xgboost <- build_XGBoost(dataset = protein_RPPA_norm)

# ROC/AUC 
protein_RPPA_ROC_xgboost <- ROC_fun(model_fit = protein_RPPA_model_xgboost)
protein_RPPA_ROC_xgboost

# evaluation index 
protein_RPPA_Evaluation_xgboost <- Evaluation_fun(model_fit = protein_RPPA_model_xgboost)
protein_RPPA_Evaluation_xgboost

# saving model 
if(!dir.exists("../../Result/classification/model")){
  dir.create("../../Result/classification/model", recursive = T)
}
save(protein_RPPA_model_xgboost, protein_RPPA_ROC_xgboost, protein_RPPA_Evaluation_xgboost, 
     file = "../../Result/classification/model/protein_RPPA_xgboost.RData")
```


### version
```{r}
sessionInfo()
```


### Reference 

1. [extreme gradient boosting](https://www.statology.org/xgboost-in-r/) 

2. [XGBoost R Tutorial](https://xgboost.readthedocs.io/en/latest/R-package/xgboostPresentation.html)

3. [Machine Learning Basics – Gradient Boosting & XGBoost](https://www.r-bloggers.com/2018/11/machine-learning-basics-gradient-boosting-xgboost/)

4. [Extreme Gradient Boosting with R](https://datascienceplus.com/extreme-gradient-boosting-with-r/)
